presto-spi/src/main/java/com/facebook/presto/spi/spiller/SpillStorageService.java

```diff
@@ -0,0 +1,33 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.spi.spiller;
+
+import java.io.IOException;
+import java.nio.file.Path;
+
+public interface SpillStorageService
+{
+    SpillFileHolder createTempFile(Path spillPath, String prefix, String suffix)
+            throws IOException;
+
+    void createDirectories(Path path)
+            throws IOException;
+
+    boolean canWrite(Path path);
+
+    double getAvailableSpacePercentage(Path path)
```

Comment:
@arhimondr : This will change the implementation of `FileSingleStreamSpillerFactory`. And in this PR the main purpose is to make `SpillStorageService` pluggable with minimal change  `FileSingleStreamSpillerFactory`.  I feel refactor `FileSingleStreamSpillerFactory` not to check space percentage can be a separate PR....

Reply:
the `StorageService` in Raptor also looks quite specific: 

https://github.com/prestodb/presto/blob/83e4afc56e97764a276dc69e8789d53731425676/presto-raptor/src/main/java/com/facebook/presto/raptor/storage/StorageService.java#L21-L38

Reply:
Yeah, it is specific to raptor. In raptor the assumption is that the storage is always a local file system, that is not true for a more generic case.